syntax = "proto2";

import "common/request.proto";

package dsnet.pbft.proto;

message RequestMessage {
    // op, timpstamp, client inside
    required dsnet.Request req = 1;
    required bytes sig = 2;
}

message ReplyMessage {
    required uint64 view = 1;
    // timestamp, client inside
    required dsnet.Request req = 2;
    required uint64 replicaid = 3;
    required bytes reply = 4;
    required bytes sig = 5;
}

message UnloggedRequestMessage {
    required dsnet.UnloggedRequest req = 1;
}

message UnloggedReplyMessage {
    required bytes reply = 1;
}

message Common {
    required uint64 view = 1;
    required uint64 seqnum = 2;
    required bytes digest = 3;
}

message PrePrepareMessage {
    required Common common = 1;
    required bytes sig = 2;
    // non-sig fields
    required RequestMessage message = 3;
}

message PrepareMessage {
    required Common common = 1;
    required uint64 replicaid = 2;
    required bytes sig = 3;
}

message CommitMessage {
    required Common common = 1;
    required uint64 replicaid = 2;
    required bytes sig = 3;
}